x/ccv/provider/keeper/punish_validator.go

```diff
@@ -0,0 +1,106 @@
+package keeper
+
+import (
+	"fmt"
+
+	errorsmod "cosmossdk.io/errors"
+	"cosmossdk.io/math"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	evidencetypes "github.com/cosmos/cosmos-sdk/x/evidence/types"
+	slashingtypes "github.com/cosmos/cosmos-sdk/x/slashing/types"
+	stakingtypes "github.com/cosmos/cosmos-sdk/x/staking/types"
+
+	"github.com/cosmos/interchain-security/v3/x/ccv/provider/types"
+)
+
+// JailAndTombstoneValidator jails and tombstones the validator with the given provider consensus address
+func (k Keeper) JailAndTombstoneValidator(ctx sdk.Context, providerAddr types.ProviderConsAddress) error {
+	validator, found := k.stakingKeeper.GetValidatorByConsAddr(ctx, providerAddr.ToSdkConsAddr())
+	if !found {
+		return errorsmod.Wrapf(slashingtypes.ErrNoValidatorForAddress, "provider consensus address: %s", providerAddr.String())
+	}
+
+	if validator.IsUnbonded() {
+		return fmt.Errorf("validator is unbonded. provider consensus address: %s", providerAddr.String())
+	}
+
+	if k.slashingKeeper.IsTombstoned(ctx, providerAddr.ToSdkConsAddr()) {
+		return fmt.Errorf("validator is tombstoned. provider consensus address: %s", providerAddr.String())
+	}
+
+	// jail validator if not already
+	if !validator.IsJailed() {
+		k.stakingKeeper.Jail(ctx, providerAddr.ToSdkConsAddr())
```

Comment:
Rereading this, I believe just this `Jail` call is enough and we do not need a `JailUntil` call. This is because [Jail](https://github.com/cosmos/cosmos-sdk/blob/v0.47.5/x/staking/keeper/slash.go) calls [jailValidator](https://github.com/cosmos/cosmos-sdk/blob/main/x/staking/keeper/val_state_change.go#L319) that removes the validator `ByPowerIndex` and therefore later on this validator would start the unbonding [here](https://github.com/cosmos/cosmos-sdk/blob/main/x/staking/keeper/val_state_change.go#L149).

Additionally, if we were to just call `Jail` here we still wouldn't be able to unjail in the future because the validator also gets tombstoned (see [here](https://github.com/cosmos/cosmos-sdk/blob/v0.47.5/x/slashing/keeper/unjail.go#L51)).

Reply:
I agree with the reasoning in general but I think it's safer to follow the [SDK evidence handler logic](https://github.com/cosmos/cosmos-sdk/blob/main/x/evidence/keeper/infraction.go#L141).